packages/webassemblyjs/src/compiler/validation/stack.js

```diff
@@ -0,0 +1,444 @@
+import { traverse } from "@webassemblyjs/ast";
+
+let errors = [];
+
+const ANY = "ANY";
+
+function checkTypes(a, b) {
+  if (a === ANY && b) {
+    return;
+  }
+
+  if (b === ANY && a) {
+    return;
+  }
+
+  if (a !== b) {
+    errors.push(`Expected type ${a} but got ${b || "none"}.`);
+  }
+}
+
+class ModuleContext {
+  constructor() {
+    this.funcs = [];
+    this.labels = [];
+
+    // Current stack frame
+    this.locals = [];
+  }
+
+  /**
+   * Functions
+   */
+  addFunction({ params: args, result }) {
+    args = args.map(arg => arg.valtype);
+    this.funcs.push({ args, result });
+  }
+
+  hasFunction(index) {
+    return this.funcs.length > index && index >= 0;
+  }
+
+  getFunction(index) {
+    return this.funcs[index];
+  }
+
+  /**
+   * Labels
+   */
+  addLabel(result) {
+    this.labels.unshift(result);
+  }
+
+  /**
+   * Locals
+   */
+  hasLocal(index) {
+    return this.locals.length > index && index >= 0;
+  }
+
+  getLocal(index) {
+    return this.locals[index];
+  }
+
+  addLocal(type) {
+    this.locals.push(type);
+  }
+}
+
+export default function validate(ast) {
+  // Module context
+  const moduleContext = new ModuleContext();
+
+  // Collect indices for funcs and globals
+  // TODO: This assumes `traverse` runs in program order which is probably wrong
+  traverse(ast, {
+    Func(path) {
+      moduleContext.addFunction(path.node);
+    }
+  });
+
+  errors = [];
+
+  // Simulate stack types throughout all function bodies
+  traverse(ast, {
+    Func(path) {
+      const expectedResult = path.node.result;
+
+      // Parameters are local variables
+      path.node.params.forEach(p => moduleContext.addLocal(p.valtype));
+
+      const resultingStack = path.node.body.reduce(
+        applyInstruction.bind(null, moduleContext),
+        path.node.params.map(arg => arg.valtype)
+      );
+
+      // Compare the two
+      let actual;
+      if (resultingStack !== false) {
+        expectedResult.map((type, i) => {
+          actual = resultingStack[resultingStack.length - 1 - i];
+          checkTypes(type, actual);
+        });
+      }
+    }
+  });
+
+  return errors;
+}
+
+function applyInstruction(moduleContext, stack, instruction) {
+  // Return was called, skip everything
+  if (stack.return) {
+    return stack;
+  }
+
+  // Workaround for node.args which sometimes does not contain instructions (i32.const, call)
```

Comment:
I've spotted this too. The shape of the AST is something we can change - should these instructions be changed for consistency?

Reply:
I made a list of possible ast modifications, Ill post them as separate issues today so we can decide individually